src/OpenTelemetry/Metrics/HistogramBuckets.cs

```diff
@@ -36,9 +40,38 @@ public class HistogramBuckets
 
         internal int IsCriticalSectionOccupied = 0;
 
+        private readonly Bucket root;
+
+        private readonly Func<double, int> findHistogramBucketIndex;
+
         internal HistogramBuckets(double[] explicitBounds)
         {
             this.ExplicitBounds = explicitBounds;
+            this.findHistogramBucketIndex = this.FindBucketIndexLinear;
+            if (explicitBounds != null && explicitBounds.Length >= DefaultHistogramCountForBinarySearch)
+            {
+                this.root = ConstructBalancedTree(explicitBounds, 0, explicitBounds.Length);
+                this.findHistogramBucketIndex = this.FindBucketIndexBinary;
+
+                static Bucket ConstructBalancedTree(double[] values, int min, int max)
```

Comment:
What's the motivation behind creating a balanced search tree? We already have the explicit bounds as a sorted array. Theoretically, the time complexity is same for "Binary Search over a sorted array" vs "Searching for an element in a balanced Binary Search Tree". This also takes extra space just to store the tree.

Reply:
This was @CodeBlanch 's idea and implementation. The benefit being that it doesn't have to mess with array and thus avoids bounds check and memory management overhead.

Reply:
Basically, it was faster this way. Pre-computes as much as it can (at the cost of some memory) so when it does the search the inner loop is tighter.

Original version:

```csharp
        private static int FindHistogramBucketIndexBinary(double[] bounds, double number)
        {
            var left = 0;
            var right = bounds.Length - 1;

            while (left <= right)
            {
                var mid = (int)Math.Floor((double)(left + right) / 2);
                if (number == bounds[mid])
                {
                    return mid;
                }
                else if (number > bounds[mid])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }

            return right + 1;
        }
```

New version:

```csharp
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal int FindBucketIndexBinary(double value)
        {
            Bucket current = this.root;

            Debug.Assert(current != null, "Bucket root was null.");

            do
            {
                if (value <= current.LowerBoundExclusive)
                {
                    current = current.Left;
                }
                else if (value > current.UpperBoundInclusive)
                {
                    current = current.Right;
                }
                else
                {
                    return current.Index;
                }
            }
            while (current != null);

            return this.ExplicitBounds.Length;
        }
```

The original has math going on, bounds checks, stuff like that. The new version just has fewer operations being executed on the CPU per call which makes it faster.

Reply:
The JIT induced additional comparison checks etc. should not make that much difference, right? This is not an end-to-end traversal of the array. We would only access the array indices a few times. (O(log(N)).

For example, for a bounds with 15 elements, we would access `bounds[mid]` at most 4 times. For 50 elements, that check would happen at most 6 times. That does not seem a lot. What's the perf difference between these two approaches?

Reply:
@utpilla Your concern here is the extra memory being taken up? I think that is totally valid and [this](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3252#discussion_r912219474) should lessen the impact.

Here is a benchmark of the two implementations above. Clock cycles add up when they are in the inner loop 😄

|                     Method | BoundCount |      Mean |    Error |   StdDev |
|--------------------------- |----------- |----------:|---------:|---------:|
| FindBucketIndexBinaryArray |       1000 | 119.71 ns | 0.134 ns | 0.118 ns |
|  FindBucketIndexBinaryTree |       1000 |  59.93 ns | 0.090 ns | 0.075 ns |

<details>

<summary>Expand for code</summary>

```csharp
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

namespace Benchmarks.Metrics
{
    public class HistogramBenchmarks
    {
        private const int MaxValue = 10000;
        private readonly Random random = new();
        private double[] bounds;
        private Bucket root;

        [Params(1000)]
        public int BoundCount { get; set; }

        [GlobalSetup]
        public void Setup()
        {
            // Evenly distribute the bound values over the range [0, MaxValue)
            this.bounds = new double[this.BoundCount];
            for (int i = 0; i < this.bounds.Length; i++)
            {
                this.bounds[i] = i * MaxValue / this.bounds.Length;
            }

            this.root = ConstructBalancedTree(this.bounds, 0, this.bounds.Length);

            static Bucket ConstructBalancedTree(double[] values, int min, int max)
            {
                if (min == max)
                {
                    return null;
                }

                int median = min + ((max - min) / 2);
                return new Bucket
                {
                    Index = median,
                    UpperBoundInclusive = values[median],
                    LowerBoundExclusive = median > 0 ? values[median - 1] : double.MinValue,
                    Left = ConstructBalancedTree(values, min, median),
                    Right = ConstructBalancedTree(values, median + 1, max),
                };
            }
        }

        [Benchmark]
        public int FindBucketIndexBinaryArray()
        {
            return this.FindBucketIndexBinaryArray(this.random.Next(MaxValue));
        }

        [Benchmark]
        public int FindBucketIndexBinaryTree()
        {
            return this.FindBucketIndexBinaryTree(this.random.Next(MaxValue));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryArray(double value)
        {
            double[] bounds = this.bounds;

            var left = 0;
            var right = bounds.Length - 1;

            while (left <= right)
            {
                var mid = (int)Math.Floor((double)(left + right) / 2);
                if (value == bounds[mid])
                {
                    return mid;
                }
                else if (value > bounds[mid])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }

            return right + 1;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryTree(double value)
        {
            Bucket current = this.root;

            Debug.Assert(current != null, "Bucket root was null.");

            do
            {
                if (value <= current.LowerBoundExclusive)
                {
                    current = current.Left;
                }
                else if (value > current.UpperBoundInclusive)
                {
                    current = current.Right;
                }
                else
                {
                    return current.Index;
                }
            }
            while (current != null);

            return this.bounds.Length;
        }

        private sealed class Bucket
        {
            public double UpperBoundInclusive { get; set; }

            public double LowerBoundExclusive { get; set; }

            public int Index { get; set; }

            public Bucket Left { get; set; }

            public Bucket Right { get; set; }
        }
    }
}
```
</details>

Reply:
I had the same question as Utkarsh. I took a look at the code in the above comment.

1. `FindBucketIndexBinaryArray` is actually slower than calling `Array.BinarySearch`: `var mid = (int)Math.Floor((double)(left + right) / 2);` is very low performance, it's not really fair to compare with it.
2. I did some improvements to the binary sort and added another method `FindBucketIndexBinaryArrayImprove`. It's not as fast as `FindBucketIndexBinaryTree`, but it's comparable now.
3. After taking into consideration the extra memory and the upfront time taken to construct the BST (both `O(n)`), I think it's debatable whether it's an improvement overall with BST.

|                            Method | BoundCount |     Mean |    Error |   StdDev |
|---------------------------------- |----------- |---------:|---------:|---------:|
|         FindBucketIndexSystemCall |       1000 | 73.85 ns | 0.456 ns | 0.427 ns |
|        FindBucketIndexBinaryArray |       1000 | 99.48 ns | 0.721 ns | 0.639 ns |
| FindBucketIndexBinaryArrayImprove |       1000 | 52.09 ns | 0.450 ns | 0.420 ns |
|         FindBucketIndexBinaryTree |       1000 | 47.11 ns | 0.211 ns | 0.176 ns |

<details>
  <summary>Click to expand!</summary>
  
  ```csharp
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

namespace Benchmarks.Metrics
{
    public class HistogramBenchmarks
    {
        private const int MaxValue = 10000;
        private readonly Random random = new();
        private double[] bounds;
        private Bucket root;

        [Params(1000)]
        public int BoundCount { get; set; }

        [GlobalSetup]
        public void Setup()
        {
            // Evenly distribute the bound values over the range [0, MaxValue)
            this.bounds = new double[this.BoundCount];
            for (int i = 0; i < this.bounds.Length; i++)
            {
                this.bounds[i] = i * MaxValue / this.bounds.Length;
            }

            this.root = ConstructBalancedTree(this.bounds, 0, this.bounds.Length);

            static Bucket ConstructBalancedTree(double[] values, int min, int max)
            {
                if (min == max)
                {
                    return null;
                }

                int median = min + ((max - min) / 2);
                return new Bucket
                {
                    Index = median,
                    UpperBoundInclusive = values[median],
                    LowerBoundExclusive = median > 0 ? values[median - 1] : double.MinValue,
                    Left = ConstructBalancedTree(values, min, median),
                    Right = ConstructBalancedTree(values, median + 1, max),
                };
            }
        }

        [Benchmark]
        public int FindBucketIndexSystemCall()
        {
            return this.FindBucketIndexSystemCall(this.random.Next(MaxValue));
        }

        [Benchmark]
        public int FindBucketIndexBinaryArray()
        {
            return this.FindBucketIndexBinaryArray(this.random.Next(MaxValue));
        }

        [Benchmark]
        public int FindBucketIndexBinaryArrayImprove()
        {
            return this.FindBucketIndexBinaryArrayImprove(this.random.Next(MaxValue));
        }

        [Benchmark]
        public int FindBucketIndexBinaryTree()
        {
            return this.FindBucketIndexBinaryTree(this.random.Next(MaxValue));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexSystemCall(double value)
        {
            double[] bounds = this.bounds;

            return Array.BinarySearch(bounds, value);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryArray(double value)
        {
            double[] bounds = this.bounds;

            var left = 0;
            var right = bounds.Length - 1;

            while (left <= right)
            {
                var mid = (int)Math.Floor((double)(left + right) / 2);
                if (value == bounds[mid])
                {
                    return mid;
                }
                else if (value > bounds[mid])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }

            return right + 1;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryArrayImprove(double value)
        {
            double[] bounds = this.bounds;

            var left = 0;
            var right = bounds.Length;

            while (left + 1 != right)
            {
                var mid = left + ((right - left) >> 1);
                if (value < bounds[mid])
                {
                    right = mid;
                }
                else
                {
                    left = mid;
                }
            }

            return left;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryTree(double value)
        {
            Bucket current = this.root;

            Debug.Assert(current != null, "Bucket root was null.");

            do
            {
                if (value <= current.LowerBoundExclusive)
                {
                    current = current.Left;
                }
                else if (value > current.UpperBoundInclusive)
                {
                    current = current.Right;
                }
                else
                {
                    return current.Index;
                }
            }
            while (current != null);

            return this.bounds.Length;
        }

        private sealed class Bucket
        {
            public double UpperBoundInclusive { get; set; }

            public double LowerBoundExclusive { get; set; }

            public int Index { get; set; }

            public Bucket Left { get; set; }

            public Bucket Right { get; set; }
        }
    }
}
  ```
</details>

Reply:
That's a great improvement @xiang17! However, I feel like it's good to use the tree approach here. Just to give you some context, this is hot path for `Histogram` metrics, so 10% difference in perf for search is not insignificant. Every `Histogram.Record` call has to search for the right bucket.

Firstly, this tree would only be constructed for cases where the boundary count >= 50 so the common regular scenarios wouldn't be affected by this.

Regarding memory usage, currently this PR would create this tree for every MetricPoint and this is a valid concern. But once we move it to AggregatorStore we would be storing only tree per Metric. That should greatly reduce the memory usage. That's planned as part of a follow-up PR. Refer to [this](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3252#discussion_r912219474).